README.md
# go-semtools

[![go-mod-version](https://img.shields.io/github/go-mod/go-version/PaulEnglert/go-semtools)](https://github.com/PaulEnglert/go-semtools)
[![tag](https://img.shields.io/github/v/tag/PaulEnglert/go-semtools)](https://github.com/PaulEnglert/go-semtools)


A toolset for working with semantic data, such as ontologies, triples and general graph data.

## Knowledge Base

The semtools define a graph scheme that consists of `Nodes` and `Statements`. Nodes can be named (ie. referenced via an Iri) or unnamed and contain data (ie. literal nodes). A Statement describes the connection between node, using a subject, predicate, object-syntax. All of these elements together are placed within a `Graph` context that provides "grouping" of statements.

Usage

    // create a new knowledge base   
    kb := NewKnowledgeBase("mybase")

    // add some statements to the base
    graph := NewNamedNode("http://mygraph")
    kb.Insert([]Statement{
        NewStatemet(
            NewNamedNode(graph.Iri() + "#Max"), NewNamedNode(graph.Iri() + "#knows"),
            NewNamedNode(graph.Iri() + "#Mara"), graph),
        NewStatemet(
            NewNamedNode(graph.Iri() + "#Mara"), NewNamedNode(graph.Iri() + "#knows"),
            NewNamedNode(graph.Iri() + "#Max"), graph),
        NewStatemet(
            NewNamedNode(graph.Iri() + "#Mara"), NewNamedNode(graph.Iri() + "#says"),
            NewLocalizedLiteral("Hi Max", "en"), graph),
    })

    // now we could query that graph for some information
    statementsAboutMara := kb.Select().
        Graph(graph). // optionally filter to specific graph
        Subject(NewNamedNode(graph.Iri() + "#Mara")).
        Results()

    for _, s := range statementsAboutMara {
        fmt.Printf("Mara %v %v\n", s.Predicate(), s.Object())
    }

## Querying

A knowledge base can either be manually worked with using the `Statements()`, or one can use the `Select()` or custom `Query` objects to work on the underlaying data. For details and examples see [Query](./query.go).

## Parsing

Knowledge Base content can be complex and need to be communicated to and from other sources. The parser component permit doing exactly that, the available parsers are currently:

* [Turtle](https://en.wikipedia.org/wiki/Turtle_(syntax)): `TurtleParser`



CHANGELOG.md
# go-semtools Changelog


## [1.0.1] - 2019-09-18
### Changed
- Fix in Equals() of Statement function when both graphs are nil


## [1.0.0] - 2019-09-17
### Refactored
- use KnowledgeBase model, simplified more similar to graph databases, rather then separated graph objects
- add advanced query functionality
- update parser to return statements, rather then full blown graphs


## [0.0.5] - 2019-09-12
### Changed
- fix literal extractor to handle empty literals ala "", or ""@de


## [0.0.4] - 2019-09-03
### Changed
- modified base directive regex to be more narrow

### Added
- more default namespaces
- add Include to Graph


## [0.0.3] - 2019-08-05
### Added
- intitial version of graph model containing Graph, Vertice, Node, NamdNode, LiteralNode, LanguageLiteralNode, TypedLiteralNode
- initial version of parser interface
- initial version of text/ttl parser (from string & to string)
- initial version of namespace structure
- initial version of rudimentary queries: Describe, DescribeMultiple, DescribeAll using NodeDescription as result container

query_test.go
package semtools


import (
	"testing"
	"github.com/sirupsen/logrus"
)


func init() {

	logrus.SetLevel(logrus.DebugLevel)

}


func TestQuery(t *testing.T) {

	kb := NewKnowledgeBase("kb")
	stmts := []Statement{
		NewStatement(NewNamedNode("max"), NewNamedNode("knows"), NewNamedNode("mara"), NewNamedNode("friends")),
		NewStatement(NewNamedNode("mara"), NewNamedNode("knows"), NewNamedNode("max"), NewNamedNode("friends")),
		NewStatement(NewNamedNode("bill"), NewNamedNode("knows"), NewNamedNode("max"), NewNamedNode("friends")),
		NewStatement(NewNamedNode("mara"), NewNamedNode("owns"), NewNamedNode("bill"), NewNamedNode("friends")),
	}
	kb.Insert(stmts)

	res := kb.Select().
		Graph(NewNamedNode("something")).Results()
	if len(res) != 0 {
		t.Errorf("Expected results to contain no statements")
	}

	res = kb.Select().
		Graph(NewNamedNode("friends")).Results()
	if len(res) != 4 {
		t.Errorf("Expected results to contain 4 statements")
	}

	res = kb.Select().
		Subject(NewNamedNode("max")).Results()
	if len(res) != 1 {
		t.Errorf("Expected results to contain 1 statements")
	}

	res = kb.Select().
		Object(NewNamedNode("max")).Results()
	if len(res) != 2 {
		t.Errorf("Expected results to contain 2 statements")
	}

	res = kb.Select().
		Predicate(NewNamedNode("knows")).Results()
	if len(res) != 3 {
		t.Errorf("Expected results to contain 3 statements")
	}

	res = kb.Select().
		Subject(NewNamedNode("max")).
		Or().
		Object(NewNamedNode("max")).Results()
	if len(res) != 3 {
		t.Errorf("Expected results to contain 3 statements")
	}

	res = kb.Select().
		Group().
			Subject(NewNamedNode("bill")).
			Object(NewNamedNode("max")).
		EndGroup().
		Or().
		Group().
			Subject(NewNamedNode("max")).
			Object(NewNamedNode("mara")).
		EndGroup().Results()
	if len(res) != 2 {
		t.Errorf("Expected results to contain 2 statements")
	}

}

namespace_test.go
package semtools

import (
	"testing"
	"github.com/sirupsen/logrus"
)


func init() {

	logrus.SetLevel(logrus.DebugLevel)

}

func TestNewNamespace(t *testing.T) {

	n := NewEmptyNamespace()
	if n == nil {
		t.Errorf("NewEmptyNamespace() fails to create new namespace")
	}

	n = NewNamespace()
	if n == nil {
		t.Errorf("NewNamespace() fails to create new namespace")
	}
	if len(n.ListKeys()) <= 0 {
		t.Errorf("NewNamespace() doesn't contain any defaults")
	}

}

func TestListKeysValues(t *testing.T) {

	n := NewEmptyNamespace()
	n.Set("test", "fullvalue")

	if len(n.ListKeys()) != 1 || n.ListKeys()[0] != "test" {
		t.Errorf("ListKeys() returns unexpected data")
	}

	if len(n.ListValues()) != 1 || n.ListValues()[0] != "fullvalue" {
		t.Errorf("ListValues() returns unexpected data")
	}

}

func TestContainsGetSet(t *testing.T) {

	n := NewEmptyNamespace()

	n.Set("test", "fullvalue")
	if n.MustGet("test") != "fullvalue" {
		t.Errorf("MustGet() returns unexpected result")
	}
	if v, ok := n.Get("test"); !ok || v != "fullvalue" {
		t.Errorf("Get() returns unexpected result")
	}
	if _, ok := n.Get("test2"); ok {
		t.Errorf("Get() returns unexpected result")
	}

	if !n.Contains("test") {
		t.Errorf("Contains() returns unexpected result")
	}
	if n.Contains("test2") {
		t.Errorf("Contains() returns unexpected result")
	}

}

func TestInclude(t *testing.T) {

	n := NewEmptyNamespace()
	n2 := NewNamespace()

	n3 := n.Include(n2)

	if len(n3.ListKeys()) <= 0 {
		t.Errorf("Include() doesn't copy over any namespaces")
	}

}

func TestCopy(t *testing.T) {

	n1 := NewNamespace()

	n2 := n1.Copy()

	if len(n2.ListKeys()) <= 0 {
		t.Errorf("Copy() doesn't create a valid copy")
	}

}


parser_ttl_test.go
package semtools

import (
	"testing"
	"fmt"
	"github.com/sirupsen/logrus"
)


func init() {

	logrus.SetLevel(logrus.DebugLevel)

}


func TestTurtleParserOptions(t *testing.T) {

	opts := &TurtleParserOptions{
		Substitute: true,
		PrettyPrint: true,
	}

	customNs := NewEmptyNamespace()
	customNs.Set("test", "http://www.test.de/test")
	opts = &TurtleParserOptions{
		Namespace: customNs,
	}

	if _, ok := opts.Namespace.Get("test"); !ok {
		t.Errorf("TurtleParserOptions not keeping Namespace")
	}

}

func TestNewTurtleParser(t *testing.T) {

	p := NewTurtleParser(nil)
	if p == nil {
		t.Errorf("NewTurtleParser(nil) fails to return a new parser")
	}

	p = NewTurtleParser(&TurtleParserOptions{Substitute: true})
	if p == nil {
		t.Errorf("NewTurtleParser(opts) fails to return a new parser")
	}

}

func TestMarshalNode(t *testing.T) {

	nn1 := NewNamedNode("http://www.test.de/test")
	nn2 := NewNamedNode("http://www.w3.org/2001/XMLSchema#string")
	rdftype := NewNamedNode("http://www.w3.org/1999/02/22-rdf-syntax-ns#type")

	ll1 := NewLocalizedLiteral("hy my name is \"Paul\"", "de")
	ll2 := NewLocalizedLiteral("ohh", "")

	tl1 := NewTypedLiteral("1", nn1)
	tl2 := NewTypedLiteral("2", nn2)

	p := NewTurtleParser(nil)
	ns := NewNamespace()


	// test
	s, err := p.marshalNode(nn1, false, ns)
	if err != nil || s != "<http://www.test.de/test>" {
		t.Errorf("marshalNode() expected '%v' but got '%v", "<http://www.test.de/test>", s)
	}
	s, err = p.marshalNode(nn1, true, ns)
	if err != nil || s != "<http://www.test.de/test>" {
		t.Errorf("marshalNode() expected '%v' but got '%v", "<http://www.test.de/test>", s)
	}
	s, err = p.marshalNode(nn2, false, ns)
	if err != nil || s != "<http://www.w3.org/2001/XMLSchema#string>" {
		t.Errorf("marshalNode() expected '%v' but got '%v", "<http://www.w3.org/2001/XMLSchema#string>", s)
	}
	s, err = p.marshalNode(nn2, true, ns)
	if err != nil || s != "xsd:string" {
		t.Errorf("marshalNode() expected '%v' but got '%v", "xsd:string", s)
	}
	s, err = p.marshalNode(ll1, false, ns)
	if err != nil || s != "\"hy my name is \\\"Paul\\\"\"@de" {
		t.Errorf("marshalNode() expected '%v' but got '%v", "\"hy my name is \\\"Paul\\\"\"@de", s)
	}
	s, err = p.marshalNode(ll2, false, ns)
	if err != nil || s != "\"ohh\"@default" {
		t.Errorf("marshalNode() expected '%v' but got '%v", "\"ohh\"@default", s)
	}
	s, err = p.marshalNode(tl1, true, ns)
	if err != nil || s != "\"1\"^^<http://www.test.de/test>" {
		t.Errorf("marshalNode() expected '%v' but got '%v", "\"1\"^^<http://www.test.de/test>", s)
	}
	s, err = p.marshalNode(tl2, true, ns)
	if err != nil || s != "\"2\"^^xsd:string" {
		t.Errorf("marshalNode() expected '%v' but got '%v", "\"2\"^^xsd:string", s)
	}
	s, err = p.marshalNode(rdftype, true, ns)
	if err != nil || s != "a" {
		t.Errorf("marshalNode() expected '%v' but got '%v'", "a", s)
	}
	s, err = p.marshalNode(rdftype, false, ns)
	if err != nil || s != "<" + rdftype.Iri() + ">" {
		t.Errorf("marshalNode() expected '%v' but got '%v'", "<" + rdftype.Iri() + ">", s)
	}

}

func TestUnmarshalNode(t *testing.T) {

	p := NewTurtleParser(nil)
	ns := NewNamespace()


	// test
	n, err := p.unmarshalNode("<http://www.test.de/test>", ns)
	if err != nil || n.(NamedNode).Iri() != "http://www.test.de/test" {
		t.Errorf("unmarshalNode() go unexpected data")
	}
	n, err = p.unmarshalNode("xsd:string", ns)
	if err != nil || n.(NamedNode).Iri() != "http://www.w3.org/2001/XMLSchema#string" {
		t.Errorf("unmarshalNode() go unexpected data")
	}
	n, err = p.unmarshalNode("\"mys\\\"astring\"", ns)
	if err != nil || n.(LocalizedLiteral).Value() != "mys\\\"astring" || n.(LocalizedLiteral).Language() != "default" {
		t.Errorf("unmarshalNode() go unexpected data")
	}
	n, err = p.unmarshalNode("\"myst\\\"@ ,ring\"@de", ns)
	if err != nil || n.(LocalizedLiteral).Value() != "myst\\\"@ ,ring" || n.(LocalizedLiteral).Language() != "de" {
		fmt.Println(n.String())
		fmt.Println(n.(LocalizedLiteral).Language())
		t.Errorf("unmarshalNode() go unexpected data")
	}
	n, err = p.unmarshalNode("\"mystring\"^^<http://www.w3.org/2001/XMLSchema#string>", ns)
	if err != nil || n.(TypedLiteral).Value() != "mystring" || n.(TypedLiteral).Type().Iri() != "http://www.w3.org/2001/XMLSchema#string" {
		t.Errorf("unmarshalNode() go unexpected data")
	}
	n, err = p.unmarshalNode("\"mystring\"^^xsd:string", ns)
	if err != nil || n.(TypedLiteral).Value() != "mystring" || n.(TypedLiteral).Type().Iri() != "http://www.w3.org/2001/XMLSchema#string" {
		t.Errorf("unmarshalNode() go unexpected data")
	}
	n, err = p.unmarshalNode("\"http://myresources:80/resources/file.py\"^^<http://www.w3.org/2001/XMLSchema#anyURI>", ns)
	if err != nil || n.(TypedLiteral).Value() != "http://myresources:80/resources/file.py" || n.(TypedLiteral).Type().Iri() != "http://www.w3.org/2001/XMLSchema#anyURI" {
		t.Errorf("unmarshalNode() go unexpected data")
	}

}

func TestMarshal(t *testing.T) {

	xsdString := NewNamedNode("http://www.w3.org/2001/XMLSchema#string")
	g := NewNamedNode("http://www.test.de/test")
	u1 := NewNamedNode("http://www.test.de/test#User1")
	u2 := NewNamedNode("http://www.test.de/test#User2")
	u3 := NewNamedNode("http://www.test.de/test#Mustermann")
	u4 := NewNamedNode("http://www.test.de/test#Max")
	u5 := NewNamedNode("http://www.test.de/test#Dirk")
	hasLN := NewNamedNode("http://www.test.de/test#hasLastName")
	hasFN := NewNamedNode("http://www.test.de/test#hasFirstName")
	says := NewNamedNode("http://www.test.de/test#says")
	l1 := NewLocalizedLiteral("my tet \"aiaiai", "en")
	t1 := NewTypedLiteral("ui a string value", xsdString)

	stmts := []Statement{
		NewStatement(u1, hasLN, u3, g),
		NewStatement(u1, hasFN, u4, g),
		NewStatement(u1, hasFN, u5, g),
		NewStatement(u1, says, l1, g),
		NewStatement(u1, says, t1, g),
		NewStatement(u2, hasLN, u3, g),
	}

	expect := `@base <http://www.test.de/test> .
<http://www.test.de/test#User1> <http://www.test.de/test#hasFirstName> <http://www.test.de/test#Dirk> , <http://www.test.de/test#Max> ; <http://www.test.de/test#hasLastName> <http://www.test.de/test#Mustermann> ; <http://www.test.de/test#says> "my tet \"aiaiai"@en , "ui a string value"^^<http://www.w3.org/2001/XMLSchema#string> .
<http://www.test.de/test#User2> <http://www.test.de/test#hasLastName> <http://www.test.de/test#Mustermann> .
`

	parser := NewTurtleParser(nil)
	
	ttl, err := parser.Marshal(stmts)
	if err != nil {
		t.Errorf("Marshal() failed")
	}
	if ttl != expect {
		fmt.Printf(expect)
		fmt.Printf(ttl)
		t.Errorf("Marshal() returned unexpected data")
	}


}

func TestMarshalWithOptions(t *testing.T) {

	xsdString := NewNamedNode("http://www.w3.org/2001/XMLSchema#string")
	g := NewNamedNode("http://www.test.de/test")
	u1 := NewNamedNode("http://www.test.de/test#User1")
	u2 := NewNamedNode("http://www.test.de/test#User2")
	u3 := NewNamedNode("http://www.test.de/test#Mustermann")
	u4 := NewNamedNode("http://www.test.de/test#Max")
	u5 := NewNamedNode("http://www.test.de/test-unsub#Dirk")
	hasLN := NewNamedNode("http://www.test.de/test#hasLastName")
	hasFN := NewNamedNode("http://www.test.de/test#hasFirstName")
	says := NewNamedNode("http://www.test.de/test#says")
	l1 := NewLocalizedLiteral("my tet \"aiaiai", "en")
	t1 := NewTypedLiteral("ui a string value", xsdString)

	stmts := []Statement{
		NewStatement(u1, hasLN, u3, g),
		NewStatement(u1, hasFN, u4, g),
		NewStatement(u1, hasFN, u5, g),
		NewStatement(u1, says, l1, g),
		NewStatement(u1, says, t1, g),
		NewStatement(u2, hasLN, u3, g),
	}

	expect := `@base <http://www.test.de/test> .
@prefix : <http://www.test.de/test#> .
@prefix bd: <http://www.bigdata.com/rdf#> .
@prefix bds: <http://www.bigdata.com/rdf/search#> .
@prefix dc: <http://purl.org/dc/elements/1.1#> .
@prefix fn: <http://www.w3.org/2005/xpath-functions#> .
@prefix foaf: <http://xmlns.com/foaf/0.1#> .
@prefix hint: <http://www.bigdata.com/queryHints#> .
@prefix owl: <http://www.w3.org/2002/07/owl#> .
@prefix rdf: <http://www.w3.org/1999/02/22-rdf-syntax-ns#> .
@prefix rdfs: <http://www.w3.org/2000/01/rdf-schema#> .
@prefix sesame: <http://www.openrdf.org/schema/sesame#> .
@prefix xsd: <http://www.w3.org/2001/XMLSchema#> .

# http://www.test.de/test#User1
:User1 
    :hasFirstName 
        :Max , 
        <http://www.test.de/test-unsub#Dirk> ; 
    :hasLastName 
        :Mustermann ; 
    :says 
        "my tet \"aiaiai"@en , 
        "ui a string value"^^xsd:string .

# http://www.test.de/test#User2
:User2 
    :hasLastName 
        :Mustermann .
`

	p := NewTurtleParser(&TurtleParserOptions{Substitute: true, PrettyPrint: true})
	
	ttl, err := p.Marshal(stmts)
	if err != nil {
		t.Errorf("Marshal() with options failed")
	}
	if ttl != expect {
		fmt.Printf(expect)
		fmt.Printf(ttl)
		t.Errorf("Marshal() returned unexpected data")
	}

}

func TestUnmarshal(t *testing.T) {
	ttl_str := `@base <http://www.test.de/test> .
				<http://www.test.de/test#User1>
					<http://www.test.de/test#hasFirstName> <http://www.test.de/test#Dirk> , <http://www.test.de/test#Max> ;
					<http://www.test.de/test#hasLastName> <http://www.test.de/test#Mustermann> ;
					<http://www.test.de/test#says> "my tet \"aiaiai"@en , "ui a string value"^^<http://www.w3.org/2001/XMLSchema#string> .
				<http://www.test.de/test#User2>
					<http://www.test.de/test#hasLastName> <http://www.test.de/test#Mustermann> ;<http://www.test.de/test#owns> "http://localhost:80/resoures/mine.py"^^<http://www.w3.org/2001/XMLSchema#anyURI> .
				:User1 a :Thing .`

	parser := NewTurtleParser(nil)
	
	stmts, err := parser.Unmarshal(ttl_str)
	if err != nil {
		t.Errorf("Unmarshal() failed: %v", err)
		return
	}
	if len(stmts) != 8 {
		t.Errorf("Unmarshal() returned unexpected number of statements")
	}

}

func TestExtractBaseIri(t *testing.T) {
	valid := `@prefix : <http://www.test.de/test#> .@base <http://www.test.de/test> .
			# http://www.test.de/test#User1
			:User1  :hasFirstName :Max .`
	nobasefull := `@prefix : <http://www.test.de/test#> .
			<http://www.test.de/test#User1>  <http://www.test.de/test#hasFirstName> <http://www.test.de/test#Max> .`
	nobasesub := `@prefix : <http://www.test.de/test#> .
			# http://www.test.de/test#User1
			:User1  :hasFirstName :Max .`
	invalid := `@prefix : <http://www.test.de/test#> .@base <http://www.test.de/test> .
			@base <http://www.test.de/test> .
			# http://www.test.de/test#User1
			:User1  :hasFirstName :Max .`
	ns := NewEmptyNamespace()
	ns.Set("", "http://www.test.de/test")

	p := NewTurtleParser(&TurtleParserOptions{RequireBaseIri: true})
	baseIri, err := p.extractBaseIri(valid, ns, false)
	if err != nil {
		t.Errorf("extractBaseIri() returned with errors")
	}
	if baseIri != "http://www.test.de/test" {
		t.Errorf("extractBaseIri() returned unexpected result")
	}

	baseIri, err = p.extractBaseIri(invalid, ns, false)
	if err == nil {
		t.Errorf("extractBaseIri() returned without errors, while it should fail")
	}

	baseIri, err = p.extractBaseIri(nobasefull, ns, false)
	if err == nil {
		t.Errorf("extractBaseIri() returned without errors, while it should fail")
	}

	baseIri, err = p.extractBaseIri(nobasefull, ns, true)
	if err != nil {
		t.Errorf("extractBaseIri() returned with errors")
	}
	if baseIri != "http://www.test.de/test#User1" {
		t.Errorf("extractBaseIri() returned unexpected data")
	}

	baseIri, err = p.extractBaseIri(nobasesub, ns, true)
	if err != nil {
		t.Errorf("extractBaseIri() returned with errors")
	}
	if baseIri != "http://www.test.de/test#User1" {
		t.Errorf("extractBaseIri() returned unexpected data")
	}

}

func TestExtractNamespace(t *testing.T) {
	valid := `@prefix : <http://www.test.de/test#> . @base <http://www.test.de/test> .
			@prefix abc: <http://www.test.de/abc#> .@prefix def: <http://www.test.de/def#> .
			# http://www.test.de/test#User1
			:User1  :hasFirstName :Max .`

	p := NewTurtleParser(nil)
	ns, err := p.extractNamespace(valid)
	if err != nil {
		t.Errorf("extractNamespace() returned with errors")
	}
	if len(ns.ListKeys()) != 3 {
		t.Errorf("extractNamespace() returned unexpected number of entries")
	}
	if v, ok := ns.Get(""); !ok || v != "http://www.test.de/test" {
		t.Errorf("extractNamespace() failed extracting namespace correctly")
	}
	if v, ok := ns.Get("abc"); !ok || v != "http://www.test.de/abc" {
		t.Errorf("extractNamespace() failed extracting namespace correctly")
	}
	if v, ok := ns.Get("def"); !ok || v != "http://www.test.de/def" {
		t.Errorf("extractNamespace() failed extracting namespace correctly")
	}
}

func TestStripComments(t *testing.T) {
	valid := `@prefix : <http://www.test.de/test#> . @base <http://www.test.de/test> .
			# http://www.test.de/test#User1
			:User1  :hasFirstName :Max .
			# http://www.test.de/test#User1
			:User1  :hasFirstName :Max .`
	expect := `@prefix : <http://www.test.de/test#> . @base <http://www.test.de/test> .
			:User1  :hasFirstName :Max .
			:User1  :hasFirstName :Max .`

	p := NewTurtleParser(nil)
	cleaned := p.stripComments(valid)
	if cleaned != expect {
		t.Errorf("stripComments() returned unexpected data")
	}
}

func TestExtractStatements(t *testing.T) {
	valid := `:User1 :hasFirstName :Max ; :hasLastName :Peter, :Dirk . <http://www.test.de/test#> rdfs:label "ta . da.s\""@de . :User2 :hasFirstName :Other .`
	expect1 := ":User1 :hasFirstName :Max ; :hasLastName :Peter, :Dirk"
	expect2 := `<http://www.test.de/test#> rdfs:label "ta . da.s\""@de`
	expect3 := ":User2 :hasFirstName :Other"

	p := NewTurtleParser(nil)
	statements := p.extractStatements(valid)
	if len(statements) != 3 {
		t.Errorf("extractStatements() returned unexpected number of statements")
	}
	if statements[0] != expect1 || statements[1] != expect2 || statements[2] != expect3 {
		t.Errorf("extractStatements() returned unexpected data in statements")
	}
}

func TestExtractSubStatements(t *testing.T) {
	valid := `:hasFirstName :Max ; :hasLastName :Peter, :Dirk`
	expect1 := ":hasFirstName :Max"
	expect2 := `:hasLastName :Peter, :Dirk`

	p := NewTurtleParser(nil)
	statements := p.extractSubStatements(valid)
	if len(statements) != 2 {
		t.Errorf("extractSubStatements() returned unexpected number of statements")
	}
	if statements[0] != expect1 || statements[1] != expect2 {
		t.Errorf("extractSubStatements() returned unexpected data in statements")
	}
}

func TestExtractObjects(t *testing.T) {
	valid := `<myuri#asds> , "wrds", "", ""@de ,"sdasds"@de,"sdasds"@de, <myuri#asds> , "sdasds"@de ,"sda\"@dasd , sds"@de , :Testx, sdas:Test, "sdasds"@de, "sdasds"^^xsd:string, "sdasds"^^<http://jdsj/asds#sads>,"http://localhost:80/resoures/mine.py"^^<http://www.w3.org/2001/XMLSchema#anyURI>`
	expect := []string{
		`"wrds"`,
		`""`,
		`""@de `,
		`"sdasds"@de`,
		`"sdasds"@de`,
		`"sdasds"@de `,
		`"sda\"@dasd , sds"@de `,
		`"sdasds"@de`,
		`"sdasds"^^xsd:string`,
		`"sdasds"^^<http://jdsj/asds#sads>`,
	 	`"http://localhost:80/resoures/mine.py"^^<http://www.w3.org/2001/XMLSchema#anyURI>`,
		`<myuri#asds> `,
		`<myuri#asds> `,
		`:Testx`,
		`sdas:Test`,
	}

	p := NewTurtleParser(nil)
	objects := p.extractObjects(valid)
	if len(objects) != len(expect) {
		t.Errorf("extractObjects() returned unexpected number of statements")
	}
	for idx, o := range objects {
		if o != expect[idx] {
			t.Errorf("extractObjects() returned unexpected data '%v' but expected '%v'", o, expect[idx])
		}
	}
	
}


query.go
package semtools

// Query creates a query structure for matching a set
// of statements either by providing the statments
// directly or by binding to a knowledge base.
type Query interface {

	// Group opens a new group. This in combination with
	// Or() allows to create more complex logic in the otherwise
	// rather simple Query interface.
	Group() Query

	// EndGroup closes an opened group.
	EndGroup() Query

	// Or combines the immediately preceeding and immediately
	// following statement with a OR-operator. Using groups
	// before and after combines the result of the groups
	// as a whole.
	Or() Query

	// Graph matches a given graph node in the statements.
	Graph(node NamedNode) Query

	// Subject matches a given subject node in the statements.
	Subject(node NamedNode) Query

	// Predicate matches a given predicate node in the statements.
	Predicate(node NamedNode) Query

	// Object matches a given object node in the statements.
	Object(node Node) Query

	// Bind binds a knowledge base to the query and will allow
	// the use of Result() and ResultIndexes()
	Bind(base KnowledgeBase) Query

	// Evaluate executes the query on the given statement
	// and returns true/false depending on if it was matched.
	Evaluate(stmt Statement) bool

	// Results executes the query on the bound knowledge base
	// returning the matched statements.
	Results() []Statement

	// ResultsFrom works like Results() just that it takes
	// the set of statements as a parameter rather then using
	// the bound knowledge base.
	ResultsFrom(stmts []Statement) []Statement

	// ResultIndexes executes the query on the bound knowledge base
	// returning the indexes of the matched statements.
	ResultIndexes() []int

	// ResultIndexesFrom works like ResultIndexes() just that it takes
	// the set of statements as a parameter rather then using
	// the bound knowledge base.
	ResultIndexesFrom(stmts []Statement) []int

}

// NewQuery creates a new query object.
func NewQuery() Query {
	return &query{
		base: nil,
		parent: nil,
		query: []matcher{func(stmt Statement) bool {return true}},
		nextOp: "and",
	}
}


// NewQueryWithParent creates a new query object using the parent.
func NewQueryWithParent(parent Query) Query {
	return &query{
		base: nil,
		parent: parent,
		query: []matcher{func(stmt Statement) bool {return true}},
		nextOp: "and",
	}
}



type matcher = func (stmt Statement) bool

type query struct {
	base KnowledgeBase
	parent Query
	query []matcher
	nextOp string
}


func (q *query) Group() Query {
	sq := NewQueryWithParent(q)
	q.add(func(stmt Statement) bool {
		return sq.Evaluate(stmt)
	})
	return sq
}

func (q *query) EndGroup() Query {
    if q.parent == nil {
        return q
    }
    return q.parent;
}

func (q *query) Or() Query {
	q.nextOp = "or"
	return q
}

func (q *query) Graph(node NamedNode) Query {
	q.add(func(stmt Statement) bool {
		if stmt.Graph() == nil {
			return node == nil
		}
		return stmt.Graph().Equals(node)  
	})
	return q
}

func (q *query) Subject(node NamedNode) Query {
	q.add(func(stmt Statement) bool {
		return stmt.Subject().Equals(node)  
	})
	return q
}

func (q *query) Predicate(node NamedNode) Query {
	q.add(func(stmt Statement) bool {
		return stmt.Predicate().Equals(node)  
	})
	return q
}

func (q *query) Object(node Node) Query {
	q.add(func(stmt Statement) bool {
		return stmt.Object().Equals(node)  
	})
	return q
}

func (q *query) Bind(base KnowledgeBase) Query {
	q.base = base
	return q
}

func (q *query) Evaluate(stmt Statement) bool {
	for _, m := range q.query {
		if !m(stmt) {
			return false
		}
	}
	return true
}

func (q *query) Results() []Statement {
	stmts := []Statement{}
	if q.base != nil {
		stmts = q.base.Statements()
	}
	return q.ResultsFrom(stmts)
}

func (q *query) ResultsFrom(stmts []Statement) []Statement {
	res := []Statement{}
	for _, stmt := range stmts {
		if q.Evaluate(stmt) {
			res = append(res, stmt)
		}
	}
	return res
}

func (q *query) ResultIndexes() []int {
	stmts := []Statement{}
	if q.base != nil {
		stmts = q.base.Statements()
	}
	return q.ResultIndexesFrom(stmts)
}

func (q *query) ResultIndexesFrom(stmts []Statement) []int {
	res := []int{}
	for idx, stmt := range stmts {
		if q.Evaluate(stmt) {
			res = append(res, idx)
		}
	}
	return res
}

func (q *query) add(qmatcher matcher) {
	op := q.nextOp
	q.nextOp = "and"

	if op == "and" {

		q.query = append(q.query, qmatcher)

	} else if op == "or" {

		lidx := len(q.query) - 1
		prev := q.query[lidx]
		q.query[lidx] = func(stmt Statement) bool {
			return prev(stmt) || qmatcher(stmt)
		}

	}

}


namespace.go
package semtools

// Namespace defines abbreviations for long uris and
// names within a graph context.
type Namespace struct {

	// ns contains the mapping between abbreviation
	// and value
	ns map[string]string
}

// NewNamespace creates a new namespace including the default
// values globally provided.
func NewNamespace() *Namespace {
	ns := NewEmptyNamespace()
	ns.Set("xsd", "http://www.w3.org/2001/XMLSchema")
	ns.Set("rdf", "http://www.w3.org/1999/02/22-rdf-syntax-ns")
	ns.Set("rdfs", "http://www.w3.org/2000/01/rdf-schema")
	ns.Set("owl", "http://www.w3.org/2002/07/owl")
	ns.Set("sesame", "http://www.openrdf.org/schema/sesame")
	ns.Set("fn", "http://www.w3.org/2005/xpath-functions")
	ns.Set("foaf", "http://xmlns.com/foaf/0.1")
	ns.Set("dc", "http://purl.org/dc/elements/1.1")
	ns.Set("hint", "http://www.bigdata.com/queryHints")
	ns.Set("bd", "http://www.bigdata.com/rdf")
	ns.Set("bds", "http://www.bigdata.com/rdf/search")
	return ns
}

// NewEmptyNamespace creates a new namespace without any
// names.
func NewEmptyNamespace() *Namespace {
	return &Namespace{ns: map[string]string{}}
}

// ListKeys provides a list of the keys (abbreviations) defined in
// the namespace.
func (ns *Namespace) ListKeys() []string {
	keys := make([]string, len(ns.ns))
	idx := 0
	for k, _ := range ns.ns {
		keys[idx] = k
		idx += 1
	}
	return keys
}

// ListValues provides a list of the values (long versions) defined
// in the namespace
func (ns *Namespace) ListValues() []string {
	values := make([]string, len(ns.ns))
	idx := 0
	for _, v := range ns.ns {
		values[idx] = v
		idx += 1
	}
	return values
}

// Contains checks if a key exists within the namespace.
func (ns *Namespace) Contains(key string) bool {
	_, ok := ns.ns[key]
	return ok
}

// Get returns the value of the key. Check the ok flag
// in order to make sure that the key was also present.
func (ns *Namespace) Get(key string) (string, bool) {
	v, ok := ns.ns[key]
	return v, ok
}

// MustGet works like Get, just assumes the Key exists.
func (ns *Namespace) MustGet(key string) string {
	v, _ := ns.Get(key)
	return v
}

// Set adds/overwrites an abbreviation with a value.
func (ns *Namespace) Set(key string, value string) {
	ns.ns[key] = value
}

// Include incorporates another namespace into itself.
func (ns *Namespace) Include(other *Namespace) *Namespace {
	copy := ns.Copy()
	for _, k := range other.ListKeys() {
		copy.Set(k, other.MustGet(k))
	}
	return copy
}

// Copy creates a copy of the namespace.
func (ns *Namespace) Copy() *Namespace {
	copy := NewNamespace()
	for _, k := range ns.ListKeys() {
		copy.Set(k, ns.MustGet(k))
	}
	return copy
}
